### ISSUE TITLE
Optimize profile export

### ISSUE THREAD

**Issue by SuperTux88** [diaspora/diaspora#7615]
After fixing diaspora/diaspora_federation#73, diaspora/diaspora_federation#75 and #7614 I was finally able to export my profile on my local computer. It took ~50 minutes and **10.8GB** RAM for sidekiq. While I don't see the time as a problem (it's an old profile, it takes some time), I think the RAM-usage is a problem. At least on my production pod I won't be able to export my profile with that RAM-usage, because the pod would be really fast out-of-memory. For my local test I restarted sidekiq before the export and the export was the only task it was doing during the 50 minutes. A production pod would also need to federate and process other task during that time. And I think that my pod will not be the only one having problems with this, especially other smaller pods which exists for some time (so they have a longer history).

The final export was 28.4MB small (gz) and after unzipping it was 105.3MB. So the export doesn't need 10.8GB. Maybe we can split the export in smaller parts and concat at the end, so we don't need to hold everything in memory?

But we need to do something before the release of 0.7.1.0 (that's why I added the `blocking-release` label), either we fix it or we should disable the export until we fixed it, otherwise this could DOS some pods.

**Comment by cmrd-senya** [comment]
I tried running profile export from rails console with my test profile from a test database using these commands:
```ruby
u = User.find_by(username: "senya")
e = Diaspora::Exporter.new(u).execute
```

I used [ruby-mass](https://github.com/archan937/ruby-mass) gem to analyze objects lifetime. What I can see is that after running export all of the `ActiveRecord` objects that was used by the serializers are still alive:

```ruby
2.4.0 :010 > Mass.index ActiveRecord::Base
 => {"User"=>[68098600, 86370220, 86818200, 86995080, 87062600], "Person"=>[69240020, 74471800, 75091100, 75112060, 81698420, 81740660, 81760940, 81801360, 82358980, 82399020, 82416540, 82431580, 82458420, 82470640, 82501180, 82524720, 82703480, 82822160, 83681720, 83701520, 83743180, 83855280, 84031020, 84043260, 84090900, 84111760, 84151040, 85712920, 85823080, 86290200, 86290660, 86291100, 86291560, 86393360, 86852080, 86945020, 87104680], "StatusMessage"=>[79843000, 81723560, 81785340, 82382280, 82439840, 82729960, 83726240, 83823460, 84071080, 85734660, 86151480, 86152000, 86152980], "Contact"=>[81762120, 81764800, 81765460, 81780520], "Aspect"=>[82378500, 82466320, 82504520, 82717860, 82718180, 82718420, 82718660, 83764080, 86772540, 86772760, 86772960, 86773160, 86994240, 86994480, 86994740, 86994920, 87154700, 87154880, 87155060, 87187240], "Profile"=>[82451500, 82489880, 82527720, 84048020, 70022261956700], "Reshare"=>[82805240, 84134660, 86887680], "Like"=>[83952320, 83952780, 83953240, 83953760, 83954260, 83954780, 83955300, 83955840, 83956420, 83956940, 83958680]}
```

I don't know what exactly references to these objects, but since they are not freed and since they are fragmented in memory large sections of memory are blocked from being freed. I guess that if we manage to unreference these objects, we'll solve this issue.

**Comment by cmrd-senya** [comment]
Okay, looks like the reason is the associations of a `User` instance. When we keep the same `User` instance during export it caches all objects we reference. We reference all user's posts, relayables and other data which  leads to all of these objects being loaded in memory at the same time. What I tried instead, is not to keep the same `User` object during the export, but to query it by id and instanciate on every `ActiveModel::Serializer#object` call.

On my local setup it works significantly better now, memory usage reaches plateau of about 600 Mb and  doesn't grow anymore.

Here is an experimental patch: https://gist.github.com/cmrd-senya/e48c6f7f75e3b7e7ad46fa79095ea285#file-export-memory-test-patch

@SuperTux88, could you please test it with your setup to see if the memory usage improves and if it improves enough?

If everything is fine, I'll create a PR for that.

**Comment by SuperTux88** [comment]
The patch is already an improvement (but not enough), memory usage (with exactly the same profile): 8.2GB.

**Comment by cmrd-senya** [comment]
@SuperTux88 could you please test another version of patch?

It is available at the [same]( https://gist.github.com/cmrd-senya/e48c6f7f75e3b7e7ad46fa79095ea285#file-export-memory-test-patch) gist.

It shows significant better results for me. For the same profile it now shows only 266 Mb of RAM usage.

I only changed it to use `find_each(batch_size: 20)` for posts queries. I didn't change anything to relayables queries. Perhaps some internal optimization already takes place for them. However it is possible that my test profile doesn't have good dataset.

**Comment by SuperTux88** [comment]
OK, it's much better now, for the same profile the RAM usage was 2.1GB now :) That can probably be improved by batch the relayables too.

But we have also another "problem" now: It took also much longer now (I canceled it after 2h for the same profile that took 50min before, it was exporting mid 2014 ... so it would have needed much longer). Because of the batching, this query (from `PostsWithActivity`) took 2.2s for every 20 posts batch:

```
Post Load (2240.0ms)  SELECT  `posts`.* FROM (SELECT DISTINCT posts.* FROM `posts` INNER JOIN `comments` ON `comments`.`commentable_id` = `posts`.`id` AND `comments`.`commentable_type` = 'Post' WHERE (`posts`.`author_id` != 1005) AND `comments`.`author_id` = 1005 UNION SELECT `posts`.* FROM `posts` INNER JOIN `likes` ON `likes`.`target_id` = `posts`.`id` AND `likes`.`positive` = 1 AND `likes`.`target_type` = 'Post' WHERE (`posts`.`author_id` != 1005) AND `likes`.`author_id` = 1005 UNION SELECT `posts`.* FROM `posts` INNER JOIN `participations` ON `participations`.`target_id` = `posts`.`id` AND `participations`.`target_type` = 'Post' WHERE (`posts`.`author_id` != 1005) AND `participations`.`author_id` = 1005 UNION SELECT `posts`.* FROM `posts` INNER JOIN `polls` ON `polls`.`status_message_id` = `posts`.`id` INNER JOIN `poll_participations` ON `poll_participations`.`poll_id` = `polls`.`id` WHERE `posts`.`type` IN ('StatusMessage') AND (`posts`.`author_id` != 1005) AND `poll_participations`.`author_id` = 1005 UNION SELECT `posts`.* FROM `posts` INNER JOIN `posts` `reshares_posts` ON `reshares_posts`.`root_guid` = `posts`.`guid` AND `reshares_posts`.`type` IN ('Reshare') WHERE (`posts`.`author_id` != 1005) AND `reshares_posts`.`author_id` = 1005) AS posts WHERE (`posts`.`id` > 349790) ORDER BY `posts`.`id` ASC LIMIT 20
```
Maybe we can select all IDs first and then batch over the IDs? Or optimize that query? Or increase the batch size?

**Comment by cmrd-senya** [comment]
I updated the [experimental patch](https://gist.github.com/cmrd-senya/e48c6f7f75e3b7e7ad46fa79095ea285) once again.

Changes:
1. I used a tricky way to serialize relayables in batches without hacking AMS, so memory footprint should go down.
2. I enlarged the batch size for posts with activity
3. I removed usage of batches for posts in `OthersDataSerializer#non_contact_authors` since it doesn't make sense there.

2 and 3 should make it run faster.

@SuperTux88, could you please test if it shows acceptable performance now or we need another improvement?

**Comment by cmrd-senya** [comment]
> Maybe we can select all IDs first and then batch over the IDs

I think it could actually work and be relatively simply. If it is too slow again, I can try this. For me this query runs fast, so I guess it is a result of database size of your pod.

**Comment by SuperTux88** [comment]
The updated code now used 1.8GB and 40 minutes with my profile 👍

**Comment by Flaburgan** [comment]
1.8 still feels high to me. What's going to happen if 3 or 4 users ask for the archive on a normal pod (~4GB memory)? I guess it is more acceptable to make a user wait 1h30 if it avoids all other users to use a really slow pod. 300 / 400MB is the maximum imo.

**Comment by cmrd-senya** [comment]
Also, I'm a little worried about the use case, when a user wants to do continuous backups. For example someone may want to request account backup every day, and this will cost much to the podmin. We should definetely do something about it.

What we can do is to restrict manual back up request to some minimal period between them, and for people who want to make backups continuously we can make some incremental backup features using API. This API will be used by some remote software, e.g. ownCloud plugin or something else which could control all of the complexity of incremental backups so that user don't have to deal with it explicitly. And for manual case we can't use this incremental approach since it's too complicated for people.

**Comment by cmrd-senya** [comment]
> 1.8 still feels high to me. What's going to happen if 3 or 4 users ask for the archive on a normal pod (~4GB memory)? I guess it is more acceptable to make a user wait 1h30 if it avoids all other users to use a really slow pod. 300 / 400MB is the maximum imo.

@Flaburgan, remember that @SuperTux88 talks about his profile which is started in 2011. Most of diaspora profile what actually take that much memory. A profile of an average user won't exceed 500Mb (or even less).

**Comment by SuperTux88** [comment]
> Also, I'm a little worried about the use case, when a user wants to do continuous backups. For example someone may want to request account backup every day, and this will cost much to the podmin.

Yes, daily backups won't be possible with the current solution (at least not with a profile as big as . But I also don't see people trying to do this currently, so I don't if we need a check for this now (and I expect people to be smart, when it takes an hour until it's finished to not click this button every day).

> What we can do is to restrict manual back up request to some minimal period between them, and for people who want to make backups continuously we can make some incremental backup features using API.

A way to create incremental backups automatically would be nice in the future. We would need to save the last IDs of every entity that we exported, so we know where to continue.

I also would like to be able to export a minimal archive (without posts and relayables, only my profile with key, contacts and aspects). That would be really fast and can also be used for an import (when the target-pod already knows the posts), full import would be better, because otherwise post subscriptions are missing, but it's still possible to import the minimal archive, because it contains all really important data. And I think a minimal-export could be added really easy and I think it should also be available via GUI.

> > 1.8 still feels high to me. What's going to happen if 3 or 4 users ask for the archive on a normal pod (~4GB memory)? I guess it is more acceptable to make a user wait 1h30 if it avoids all other users to use a really slow pod. 300 / 400MB is the maximum imo.
>
> @Flaburgan, remember that @SuperTux88 talks about his profile which is started in 2011. Most of diaspora profiles won't actually take that much memory. I guess a profile of an average user won't exceed 500Mb (or even less).

Yes, my profile is really big (@Waithamai is bigger, but most profiles would be smaller), so I think I'm happy when my profile exports with ~1GB (it's still ~150-200 MB base footprint of sidekiq, so actual memory used for the export would be ~800MB). Average profiles would use less memory, so I think we should be OK for now with that (there is still space for improvements, but it's not a release blocker anymore then).

We should maybe add a check, that only one export run at once, so when a export is already running, a new export is scheduled to run later. We can make this configurable, so bigger pods (with more users and more resources) can increase this limit, otherwise queued exports could pileup.

**Comment by cmrd-senya** [comment]
> I also would like to be able to export a minimal archive (without posts and relayables, only my profile with key, contacts and aspects).

Maybe then we can make minimal profile available on-demand, in a request-response basis, avoiding sidekiq? So that there is not only difference in size, but in usecases. Full backup for the case when we want to save all of our data and small archive for the case when we want to access our basic data from another application. For example we can evetually implement something like nomadic identities from Hubzilla this way. Also it makes sense to implement the "small archive" request as a part of our future API. Actually, I think it would be smart to add full archive export request as an API method as well.

So when you're changing your pod you're expected to use full archive and minimal archive is used for some other purposes, when you don't need all of this heavy stuff.

**Comment by SuperTux88** [comment]
> Maybe then we can make minimal profile available on-demand, in a request-response cycle, avoiding sidekiq?

No, it still takes a few seconds to collect/render it, because it's still everything in there needed for an import (including all contacts), just not the posts/comments. And I don't want long running requests in the frontend. And I also don't want to have this as an API. When we add an API there will be routes to request the data you need, I don't know for which usecase you need an API route to get all data at once.

> So when you're changing your pod you're expected to use full archive and minimal archive is used for some other purposes, when you don't need all of this heavy stuff.

Both archives can be used for the import. I see the usecase for a minimal export more like an emergency export, for example when the big export fails because of unexpected issues (like those I fixed in #7614), and your podmin doesn't react to fix them, you would still be able to migrate with the minimal export. And since it isn't much work (because it's similar to the current export, just don't add some data to the json), so I think it can be added easily?

> Actually, I think it would be smart to add full archive export request as an API method as well.

👍

**Comment by cmrd-senya** [comment]
@SuperTux88, can you please test another version of [my patch](https://gist.github.com/cmrd-senya/e48c6f7f75e3b7e7ad46fa79095ea285)?

I noticed, that AMS called `serializable_data` method on the root serializers which actually performed serialization for the second time. This also was the cause of memory bloat on "others posts" since it didn't went in batches there. I simple overrode the method with empty hash which looks like it helped.

**Comment by SuperTux88** [comment]
Closed via #7627
